doc/design/clustered-etcd-alert-store/design.md

```diff
@@ -0,0 +1,150 @@
+# Clustered Etcd Alert Store
+
+Type: Design Document (and PR)
+
+Date: 2019-09-20
+
+Author: David C Wang <dcwangmit01@gmail.com>
+
+## Status Quo
+
+The current Alertmanager clustering implementation will allow multiple
+instances of Alertmanager to sync recently-updated notification log entries and
+alert silences via the gossip protocol.  The syncing of alert state is not yet
+implemented.
+
+Currently, Alertmanager alert state is entirely stored in process memory.  It
+is lost upon restart of Alertmanager.  This, in conjunction with the fact that
+Alertmanager alert state is not synced with clustered peers presents several
+problems:
+
+* Alert emitters must send alerts to every single Alertmanager instance in the
+  cluster.  If some Alertmanager instances receive alerts that other instances
+  do not, their state will drift apart.
+
+  * Typically, the alert emitter is Prometheus which uses service discovery to
+    locate all Alertmanager instances, and then sends alerts to each individual
+    target.  Custom alert emitters may not have that service discovery
+    capability and must be explicity configured.
+
+* Because the alert must be sent to every Alertmanager instance, this precludes
+  a typical HA deployment pattern of having clients (i.e. alert emitters) send
+  alerts through a Load Balancer that fronts several Alertmanager instances.
+  This pattern is typical in Kubernetes.
+
+* Restarting Alertmanager instances or increasing the replicaCount for the
+  Alertmanager cluster will result in the new members having very different
+  alert states than current members.
```

Comment:
Can you add more details in which scenario scaling the alertmanager cluster is not a irregular process?

Reply:
Scaling the cluster shouldn't be often.  The point I was trying to make is that there are multiple ways of getting the Alertmanagers within the same cluster to have different internal alert states, because alert state is not persisted.  Scaling up or down the cluster is one.  Doing a kubernetes rolling update of the Alertmanager Statefulset is another.  Chaoskube kills pods randomly.  Cluster-autoscaler moves pods off of nodes before downscaling the number of nodes, etc.

Reply:
If it's entirely inside a k8 cluster, then SD can handle it. If it's going across clusters then you probably have well-known DNS names which will be used.